indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java

```diff
@@ -1316,22 +1304,76 @@ public ImmutableWorkerInfo apply(ZkWorker input)
   {
     return ImmutableList.copyOf(
         Collections2.transform(
-          workers,
-          new Function<ZkWorker, Worker>()
-          {
-            @Override
-            public Worker apply(ZkWorker input)
+            workers,
+            new Function<ZkWorker, Worker>()
             {
-              return input.getWorker();
+              @Override
+              public Worker apply(ZkWorker input)
+              {
+                return input.getWorker();
+              }
             }
-          }
         )
     );
   }
 
   public Collection<ImmutableWorkerInfo> getBlackListedWorkers()
   {
-    return getImmutableWorkerFromZK(blackListedWorkers);
+    synchronized (blackListedWorkers) {
+      return getImmutableWorkerFromZK(blackListedWorkers);
+    }
+  }
+
+  private boolean shouldRemoveNodeFromBlackList(ZkWorker zkWorker)
+  {
+    if (blackListedWorkers.size() > zkWorkers.size() * (config.getMaxPercentageBlacklistWorkers() / 100.0)) {
```

Comment:
I'm not sure if calling `size()` here is going to work properly given that this method is called from within an iteration loop that may also mutate the collection. Can you please confirm that and work around it if it does end up being a problem?